### ISSUE TITLE
Conditional Compilation and XmlSerialization

### ISSUE THREAD

**Issue by robertmuehsig** [dotnet/runtime#103880]
### Description

Hi everyone,
I'm not sure if this is the right repo or not, but I have an issue porting an application from .NET 6.0 to .NET 8.0. 
I'm not even sure if the "root" problem has something to do with the XmlSerializer or not.

__Our scenario:__

Our solution contains two main applications: 
- Client, which still runs on .NET Framework 4.8
- Server, which now runs on .NET 6 and we try to migrate it to .NET 8

Both applications share some common types and the actual solution is "large" (~100 Projects, multiple web apps (e.g. WebApi, Admin App etc.)).

We try to stay up to date and want to use new languages features, that's why we have a project that contains "BCL-Types" like this:

```
#if !NET7_0_OR_GREATER

namespace System.Runtime.CompilerServices
{
    /// <summary>Specifies that a type has required members or that a member is required.</summary>
    [AttributeUsage(AttributeTargets.Class | AttributeTargets.Struct | AttributeTargets.Field | AttributeTargets.Property, AllowMultiple = false, Inherited = false)]
// #if SYSTEM_PRIVATE_CORELIB
    public
// #else
//     internal
// #endif
        sealed class RequiredMemberAttribute : Attribute
    { }
}

#endif
```

This way we can use something like this in the Common area and don't bother if it is used from .NET 4.8 or not:

```
public required string Name { get; set; }
```

We use `XmlSerialization` for configuration stuff in our application and during the .NET 8 migration we found this RunTime Exception:

`Could not load type 'System.Runtime.CompilerServices.RequiredMemberAttribute' from assembly 'BclTest.BclTypes, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null'.`

<details><summary>Details</summary>
<p>

```
System.TypeLoadException
  HResult=0x80131522
  Message=Could not load type 'System.Runtime.CompilerServices.RequiredMemberAttribute' from assembly 'BclTest.BclTypes, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null'.
  Source=System.Private.CoreLib
  StackTrace:
   at System.ModuleHandle.ResolveType(QCallModule module, Int32 typeToken, IntPtr* typeInstArgs, Int32 typeInstCount, IntPtr* methodInstArgs, Int32 methodInstCount, ObjectHandleOnStack type)
   at System.ModuleHandle.ResolveTypeHandle(Int32 typeToken, RuntimeTypeHandle[] typeInstantiationContext, RuntimeTypeHandle[] methodInstantiationContext)
   at System.Reflection.RuntimeModule.ResolveType(Int32 metadataToken, Type[] genericTypeArguments, Type[] genericMethodArguments)
   at System.Reflection.CustomAttribute.FilterCustomAttributeRecord(MetadataToken caCtorToken, MetadataImport& scope, RuntimeModule decoratedModule, MetadataToken decoratedToken, RuntimeType attributeFilterType, Boolean mustBeInheritable, ListBuilder`1& derivedAttributes, RuntimeType& attributeType, IRuntimeMethodInfo& ctorWithParameters, Boolean& isVarArg)
   at System.Reflection.CustomAttribute.AddCustomAttributes(ListBuilder`1& attributes, RuntimeModule decoratedModule, Int32 decoratedMetadataToken, RuntimeType attributeFilterType, Boolean mustBeInheritable, ListBuilder`1 derivedAttributes)
   at System.Reflection.CustomAttribute.GetCustomAttributes(RuntimeModule decoratedModule, Int32 decoratedMetadataToken, Int32 pcaCount, RuntimeType attributeFilterType)
   at System.Reflection.CustomAttribute.GetCustomAttributes(RuntimeType type, RuntimeType caType, Boolean inherit)
   at System.Xml.Serialization.TempAssembly.LoadGeneratedAssembly(Type type, String defaultNamespace, XmlSerializerImplementation& contract)
   at System.Xml.Serialization.XmlSerializer..ctor(Type type, String defaultNamespace)
   at CommonTypes.Person.DeserializeFromXml(String xml) in C:\Users\muehsig\source\repos\BclTest.BclTypes\CommonTypes\Person.cs:line 13
   at TestProject1.UnitTest1.TestMethod1() in C:\Users\muehsig\source\repos\BclTest.BclTypes\TestProject1\UnitTest1.cs:line 33
   at System.RuntimeMethodHandle.InvokeMethod(Object target, Void** arguments, Signature sig, Boolean isConstructor)
   at System.Reflection.MethodBaseInvoker.InvokeWithNoArgs(Object obj, BindingFlags invokeAttr)
```
</p>
</details> 

It seems, that the XmlSerializer uses the "wrong" type during runtime. 

### Reproduction Steps

I made a small test repo [here](https://github.com/robertmuehsig/TestOldAndNew).

The smallest repo would be something like this:

```
- Common Lib -> TargetFramework "netstandard2.0"
- .NET 4.8 Console App -> TargetFramework "48"
- .NET 8 Console App -> TargetFramework "net8.0"
- BCL-Type Lib -> TargetFramework "netstandard2.0,net8.0"
- Unit Test Project -> TargetFramework "net8.0"
```

In the Common Lib is a class like this:

```
    public class Person
    {
        public required string Name { get; set; }

        public static Person DeserializeFromXml(string xml)
        {
            XmlSerializer serializer = new XmlSerializer(typeof(Person));
            using (StringReader reader = new StringReader(xml))
            {
                return (Person)serializer.Deserialize(reader);
            }
        }
    }
```

In the .NET 8 Console App is a class like this:

```
    internal class FoobarJob
    {
        public void Run([NotNullWhen(returnValue: false)] out string? errorMessage)
        {
            errorMessage = null;
            Console.WriteLine("Running FoobarJob");
        }
    }
```

In the Unit Test we have this:

```
    [TestClass]
    public class UnitTest1
    {
        [TestMethod]
        public void TestMethod1()
        {
            // works
            Person person = new Person() { Name = "Test" };

            string xml = @"<Person><Name>John Doe</Name></Person>";
            Person personX = Person.DeserializeFromXml(xml);
            Console.WriteLine(personX.Name);
        }
    }
```

### Expected behavior

The Common type can be consumed from the .NET 4.8 client and the .NET 8 application and we can use modern language features in the .NET 8 application.

### Actual behavior

__With__ the TargetFrameworks in the BCL class like this "netstandard2.0,net8.0" the runtime error occures:

`Could not load type 'System.Runtime.CompilerServices.RequiredMemberAttribute' from assembly 'BclTest.BclTypes, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null'.`

I thought, that the TargetFramework might be "wrong" and I just use "netstandard2.0" in the BCL-Type lib, but then I get this compiler error:

```
FoobarJob.cs(12,26,12,37): error CS0433: The type 'NotNullWhenAttribute' exists in both 'BclTest.BclTypes, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null' and 'System.Runtime, Version=8.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a'
3>Done building project "NetEightApp.csproj" -- FAILED.
```

### Regression?

If I use "net6.0" instead of "net8.0" the issue is resolved, but the reason for this is simply this condition: `#if !NET7_0_OR_GREATER`

### Configuration

.NET Framework 4.8 & .NET 8

**Comment by huoyaoyuan** [comment]
This is because binary compatibility is broken between target frameworks.

To achieve compatibility, you need to redirect the custom defined type to BCL defined type on higher versions, like this:

```csharp
#if !NET7_0_OR_GREATER
// definition of RequiredMemberAttribute
#else
[TypeForwardedTo(typeof(RequiredMemberAttribute))]
#endif
```

**Comment by robertmuehsig** [comment]
@huoyaoyuan Thats a good hint, but 

```
#if !NET7_0_OR_GREATER
namespace System.Runtime.CompilerServices
{
    ...
}
#else
[TypeForwardedTo(typeof(System.Runtime.CompilerServices.RequiredMemberAttribute))]
#endif
```

Results in  ` error CS0116: A namespace cannot directly contain members such as fields, methods or statements`

**Comment by huoyaoyuan** [comment]
Are you typing correctly? The content under `#if !NET7_0_OR_GREATER` should behave as-is.

**Comment by robertmuehsig** [comment]
Ok, this compiles:

```
#if !NET7_0_OR_GREATER

namespace System.Runtime.CompilerServices
{
    /// <summary>Specifies that a type has required members or that a member is required.</summary>
    [AttributeUsage(AttributeTargets.Class | AttributeTargets.Struct | AttributeTargets.Field | AttributeTargets.Property, AllowMultiple = false, Inherited = false)]
// #if SYSTEM_PRIVATE_CORELIB
    public
// #else
//     internal
// #endif
        sealed class RequiredMemberAttribute : Attribute
    { }
}
#else
[assembly:System.Runtime.CompilerServices.TypeForwardedTo(typeof(System.Runtime.CompilerServices.RequiredMemberAttribute))]
#endif
```

But the runtime error still appears:

```
System.TypeLoadException
  HResult=0x80131522
  Message=Could not load type 'System.Runtime.CompilerServices.CompilerFeatureRequiredAttribute' from assembly 'BclTest.BclTypes, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null'.
  Source=System.Private.CoreLib
  ...
```

**Comment by robertmuehsig** [comment]
Found a workaround, but I'm still confused why.

If I add `net8.0` in the TargetFrameworks of the __Common__ Lib, then it works:


```
CommonTypes.csproj:

<TargetFrameworks>netstandard2.0;net8.0</TargetFrameworks>
```

The TypeForwarding is not needed in this case.

But I wonder if this is by design or not?

**Comment by huoyaoyuan** [comment]
> If I add `net8.0` in the TargetFrameworks of the **Common** Lib, then it works:

In this case, the `net8.0` version of unit test will use `net8.0` version of **Common**, which uses the BCL version of attributes. It's expected to work in this case.

**Comment by StephenMolloy** [comment]
This issue is not fundamentally about `XmlSerializer` itself. The runtime error and compiler conflicts described are more about how multi-targeting, netstandard compatibility, and type resolution work across different frameworks. When you multi-target a library (like netstandard2.0 and net8.0), the consuming project will pick the closest compatible target. I asked Copilot for help in generating a clear explaination - and got a dissertation back. But I think you'll get the idea pretty quickly in the beginning of the novel I'm about to paste. ;)

---

## Project Setup Recap

- **Common Lib**: Targets netstandard2.0.  
  Contains shared types like `Person`, which uses `required` properties (thus `RequiredMemberAttribute`).

- **BCL-Type Lib**: Targets netstandard2.0 and net8.0.  
  Implements missing BCL attributes when not present, via conditional compilation using `#if !NET7_0_OR_GREATER`.

- **Client App**: Targets .NET Framework 4.8.

- **Server App**: Targets .NET 8.0.

- **Unit Test Project**: Targets .NET 8.0.

## 1. Multi-Targeting and Type Implementation

- In **BCL-Type Lib**, you conditionally declare types like `RequiredMemberAttribute` only when targeting frameworks less than .NET 7.

- When you build BCL-Type Lib for netstandard2.0, these types are included.  
  When you build for net8.0, they are NOT included, because .NET 8 already has these types in its own BCL.

## 2. Dependency Resolution

### What Happens When Consuming Apps Reference Common and BCL-Types?

- **netstandard2.0** is meant to be shared:  
  - .NET 4.8 uses your custom attribute definitions from BCL-Type Lib.
  - .NET 8.0, referencing BCL-Type Lib built for net8.0, expects to use the built-in framework versions of those attributes.

### **But here’s the catch:**

#### When you reference BCL-Type Lib from Common and consuming apps...

- **net8.0 consuming app** gets BCL-Type Lib’s net8.0 assembly (no custom attributes, relies on framework types).
- **Common Lib** was compiled against netstandard2.0, and so references BCL-Type Lib’s netstandard2.0 assembly (with your custom attributes included).

#### At runtime:

- The .NET 8 app loads both Common (netstandard2.0) and BCL-Type Lib (net8.0).
- But if the assembly references to BCL-Type Lib resolve to the netstandard2.0 version, the runtime sees duplicate types:
    - Your custom `RequiredMemberAttribute` exists in BCL-Type Lib netstandard2.0.
    - The real `RequiredMemberAttribute` exists in System.Runtime.CompilerServices in .NET 8’s BCL.

#### This leads to two possible errors:

1. **TypeLoadException at runtime**  
   If the runtime tries to load `RequiredMemberAttribute` from your assembly when it should be using the one from the framework, you get an error like:
   ```
   Could not load type 'System.Runtime.CompilerServices.RequiredMemberAttribute' from assembly 'BclTest.BclTypes...'
   ```
   This happens because the runtime finds the attribute defined in your library, but it's not the one the BCL expects (it’s a duplicate, potentially with different accessibility or implementation).

2. **Compiler ambiguity error**  
   If you only target netstandard2.0 in BCL-Type Lib, but consume it from a net8.0 app, both your library and the BCL declare the same types. The compiler sees two candidates:
   ```
   The type 'NotNullWhenAttribute' exists in both 'BclTest.BclTypes...' and 'System.Runtime...'
   ```
   This is a classic diamond dependency problem: two assemblies define the same type in the same namespace.

## 3. How Multi-Target Dependency Resolution Works

- When you reference a multi-targeted library, the consuming project picks the best match for its target framework:
    - net8.0 app resolves to BCL-Type Lib’s net8.0 build.
    - net4.8 app resolves to BCL-Type Lib’s netstandard2.0 build.

- However, if you share types via a netstandard2.0 Common library, any downstream consumer might transitively pull in netstandard2.0 versions of dependencies (like BCL-Type Lib), even if a direct reference would prefer the net8.0 build.

- This is called “dependency unification” or “assembly resolution.” If the dependency graph causes both netstandard2.0 and net8.0 versions to be referenced, resolution can be ambiguous or incorrect—especially if types with the same name exist in both.

## 4. Why XmlSerializer Triggers the Problem

- XmlSerializer uses reflection to inspect custom attributes at runtime.
- If there are mismatches in type identity (two distinct `RequiredMemberAttribute` types with the same name/namespace), reflection can’t resolve which one to use, leading to TypeLoadException.

---

**In summary:**  
The errors stem from how assembly and type resolution works in multi-targeted projects, especially when you define types that newer frameworks now include. The runtime and compiler confusion is due to duplicate types being present in both your library and the BCL, and mismatches in which assembly is actually loaded for a given dependency.
